### PR TITLE
Remove temporary image and delta layer files if 'finish' wasn't called

### PR DESCRIPTION
**Pull request by alexstanovoy**
The fix of https://github.com/neondatabase/neon/issues/2650

Contains the actual fix (drop implementations) and two regression tests for the image layer files, and delta layer files.

### CODE REVIEW

**Comment by alexstanovoy** [comment]
This still requires a second test for delta layers. Also, the first test isn't actually testing anything: the files aren't created even when compaction is successful. I'm investigating and fixing that.

File: pageserver/src/tenant/timeline.rs [hunk 1]
```diff
@@ -1541,7 +1541,9 @@ impl Timeline {
                     lsn,
                 )?;
 
+                info!("blabla for");
```
Comment (SomeoneToIgnore) [inline pageserver/src/tenant/timeline.rs:1544]: NIT: Simpler to mark such PRs as drafts if you have debugging enabled :slightly_smiling_face:

File: test_runner/regress/test_image_layers.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+from fixtures.neon_fixtures import NeonEnv, wait_for_last_flush_lsn
+
+
+def test_image_layer_fail_before_finish(neon_simple_env: NeonEnv):
+    pageserver_http = neon_simple_env.pageserver.http_client()
+    pageserver_http.configure_failpoints(("image-layer-fail-before-finish", "return"))
+
+    image_files = []
+
+    with pytest.raises(Exception):
```
Comment (SomeoneToIgnore) [inline test_runner/regress/test_image_layers.py:18]: What exactly fails here? I would not expect tenant and branch creation to fail in the test, only the compaction method.
So we might want to move this `with` to the smallest possible scope around the exception triggering method.

File: test_runner/regress/test_image_layers.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                "CREATE TABLE foo (t text) WITH (autovacuum_enabled = off)",
+                """INSERT INTO foo
+            SELECT 'long string to consume some space' || g
+            FROM generate_series(1, 100000) g""",
+            ]
+        )
+        wait_for_last_flush_lsn(neon_simple_env, pg, neon_simple_env.initial_tenant, new_timeline_id)
+        pageserver_http.timeline_compact(tenant_id, new_timeline_id)
+
+        image_files = [path for path in neon_simple_env.timeline_dir(tenant_id, new_timeline_id).iterdir()]
```
Comment (SomeoneToIgnore) [inline test_runner/regress/test_image_layers.py:43]: Those are called "layer files" generally, there are delta and image layers on disk, so image_files is a bit confusing in the context.

I wonder, would it make sense to first create a timeline without the failpoint enabled, then enable it and trigger compaction: this way, we're creating some initial layers and ensure that failed compaction does not leave behind any tmp files.

**Comment by SomeoneToIgnore** [comment]
With the current settings, we fail to create initial checkpoint thus the timeline entirely, if I'm not mistaken.

* In order to test delta layer creation, you would need to enable failpoint after the branch creation: this way you'll get some delta layers and will be able to generate another one with the failopint enabled. Reducing the `checkpoint_distance` parameter would help to speed up the test.

* Image layers are produced after the compaction process, out of `compaction_threshold` delta layer files on compaction call after enough delta layers is accumulated. That would require enabling failpoints only before the compaction is triggered (note `compaction_period` is set to 10 mintues in your test, so technically that could interfere).

File: pageserver/src/tenant/delta_layer.rs [hunk 4]
```diff
@@ -768,6 +768,57 @@ impl DeltaLayerWriter {
...
+    pub fn size(&self) -> u64 {
+        self.inner.as_ref().unwrap().size()
+    }
+
+    pub fn finish(mut self, key_end: Key) -> Result<DeltaLayer> {
+        self.inner.take().unwrap().finish(key_end)
+    }
+}
+
+impl Drop for DeltaLayerWriter {
```
Comment (hlinnaka) [inline pageserver/src/tenant/delta_layer.rs:811]: Please add a comment here, to explain what this is for. (To remove the incomplete file if the caller drops the writer without calling `finish`)

File: test_runner/regress/test_layer_writers_fail.py [hunk 5]
```diff
@@ -29,17 +30,55 @@ def test_image_layer_writer_fail_before_finish(neon_simple_env: NeonEnv):
...
     )
 
-    wait_for_last_flush_lsn(env, pg, tenant_id, timeline_id)
-    pageserver_http.timeline_checkpoint(tenant_id, timeline_id)
+    pageserver_http.configure_failpoints(("image-layer-writer-fail-before-finish", "return"))
+    with pytest.raises(Exception, match="image-layer-writer-fail-before-finish"):
+        pageserver_http.timeline_checkpoint(tenant_id, timeline_id)
 
-    old_layer_files = [path for path in env.timeline_dir(tenant_id, timeline_id).iterdir()]
+    is_temp_path = lambda file: str(file)[-7:] == '___temp'
```
Comment (SomeoneToIgnore) [inline test_runner/regress/test_layer_writers_fail.py]: This particular prefix is pageserver's knowledge: https://github.com/neondatabase/neon/blob/259a5f356e036b5cf274cd2222884434dd8c55f0/pageserver/src/lib.rs#L128 so we might want to put it as a const in `neon_fixtures.py::NeonPageserver`.

One idea for future would be to check every timeline and tenant creation failure test for absence of temp files, that constant would help to do it later.

File: pageserver/src/tenant/delta_layer.rs [hunk 6]
```diff
@@ -768,6 +768,101 @@ impl DeltaLayerWriter {
...
+/// # Note
+///
+/// As described in https://github.com/neondatabase/neon/issues/2650, it's
+/// possible for the writer to drop before `finish` is actually called. So this
+/// could lead to odd temporary files in the directory, exhausting disk space.
+/// This structure wraps `DeltaLayerWriterInner` and also contains `Drop`
+/// implementation that cleans up the temporary file in failure. It's not
+/// possible to do this directly in `DeltaLayerWriterInner` since `finish` moves
+/// out some fields, making it impossible to implement `Drop`.
+///
```
Comment (SomeoneToIgnore) [inline pageserver/src/tenant/delta_layer.rs:791]:
```suggestion
#[must_use]
```

We can add this to both builders, to help reduce programming errors, that cause dropping the builder.

**Review by SomeoneToIgnore — APPROVED** [review]
LGTM, left a few NITs

File: test_runner/regress/test_layer_writers_fail.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    assert (
+        len(new_temp_layer_files) == 0
+    ), "pageserver should clean its temporary new image layer files on failure"
+
+
+def test_delta_layer_writer_fail_before_finish(neon_simple_env: NeonEnv):
+    env = neon_simple_env
+    pageserver_http = env.pageserver.http_client()
+
+    tenant_id, _ = env.neon_cli.create_tenant(
```
Comment (SomeoneToIgnore) [inline test_runner/regress/test_layer_writers_fail.py:57]: Both tests could run on `main` branch  (it has id `_` here) with the same effect, but less code.
Your test case it not really related to branching, so might be simpler to remove it.

**Comment by alexstanovoy** [comment]
Created an issue https://github.com/neondatabase/neon/issues/2703. I'll merge the fix and the tests, but for some time they'll be skipped since they're flaky because of background compaction problems.

**Pull request merged**